packages/material-ui/src/Menu/Menu.js

```diff
@@ -94,10 +152,38 @@ const Menu = React.forwardRef(function Menu(props, ref) {
       <MenuList
         data-mui-test="Menu"
         onKeyDown={handleListKeyDown}
-        {...MenuListProps}
         actions={menuListActionsRef}
+        autoFocus={menuListAutoFocus}
+        {...MenuListProps}
       >
-        {children}
+        {React.Children.map(children, (child, index) => {
+          if (!React.isValidElement(child)) {
+            return null;
+          }
+
+          if (
+            index === tabIndex0Index ||
+            index === autoFocusIndex ||
+            index === contentAnchorElIndex
+          ) {
+            const newChildProps = {};
+            if (index === autoFocusIndex) {
+              newChildProps.autoFocus = true;
+            }
+            if (index === tabIndex0Index) {
+              newChildProps.tabIndex = 0;
+            }
+            if (index === contentAnchorElIndex) {
+              newChildProps.ref = refArg => {
+                // not StrictMode ready
```

Comment:
That is StrictMode ready though. All the user has to do is not use class components. If they just use material-ui components there shouldn't be any strict mode warning.

Reply:
@eps1lon I would like to get your thoughts on a different approach for setting refs. I was considering creating a `setDOMNodeRef` helper function that would check whether the `value` is a DOM node based on a heuristic of having a `nodeType` property. It would then only call `findDOMNode` if it isn't a DOM node and also provide a warning in that case. This would provide a clear path for eliminating the `findDOMNode` calls in v5.

Reply:
I'm usually very careful when it comes to using react internals. From my experience react libraries relying on them are prone to bugs between minor react releases. 

What problem do you want to solve? Calling `findDOMNode` on a host instance has virtually no cost: `findDOMNode` returns early with the input value if the input is either nullish or is a host instance. 

I wouldn't introduce a warning if we call it on a class component. The user can already opt-in to a warning with `React.StrictMode`.

Reply:
I was unclear about the extent of deprecation and StrictMode warning cases for `findDOMNode` and was thinking the goal was to eventually eliminate the calls to `findDOMNode`.

> I'm usually very careful when it comes to using react internals.

I'm not sure what react internals you are referring to unless you mean relying on class instances not having a [`nodeType`](https://developer.mozilla.org/en-US/docs/Web/API/Node/nodeType) property. It would seem like a bizarre choice for React to add such a property to class instances in the future since it would make them look more like DOM nodes and be a breaking change for anyone using `nodeType` to recognize DOM nodes.

> What problem do you want to solve?

If calling `findDOMNode` on DOM nodes is fine for the long-term (i.e. won't produce any warnings in StrictMode and is expected to stay that way), then there is no problem to solve (which is why I wanted to get your thoughts before doing it). This also helps clarify for me the exact purpose/meaning of the `// StrictMode ready` comments.

Reply:
> If calling findDOMNode on DOM nodes is fine for the long-term (i.e. won't produce any warnings in StrictMode and is expected to stay that way)

For sure. I still haven't heard back from facebook/react#15091 so this is still a bit open.

It is somewhat public how to determine if a component is a class component: https://overreacted.io/how-does-react-tell-a-class-from-a-function/

I'm not sure if there are documented approaches for host instances.

Reply:
Can't we just use an `instanceOf` check? Seems like a much safer option.

Reply:
Yes, I think we could use `instanceof Node` or `instanceof Element`. Though I think I'm strongly leaning towards your initial reaction that this introduces an undue burden on users who aren't trying to use StrictMode. In order to use this broadly, we would be saying that any custom components you specify as `component` props or wrapper components (in the case of things like `MenuItem`) need to be well-behaved as far as forwarding refs.

A lot of this would not be easily within the user's control. For instance, react-router's [`Link`](https://github.com/ReactTraining/react-router/blob/master/packages/react-router-dom/modules/Link.js) (which we specify as the `component` prop in our application on many buttons and navigation list items) is currently class-based without any ref forwarding. We probably need to continue leveraging `findDOMNode` until some alternative is provided by React (I remember reading about possible future support for a `ref` on a fragment returning the DOM nodes).

Reply:
react-router (or rather hocs) is one of the main concerns why I still want to use findDOMNode. Most hocs are either implemented as a class component or a function component with React.forwardRef. But those that are implemented as a class based component usually have some additional `innerRef` or `forwardedRef` which means you can patch those in your own code: https://next--material-ui.netlify.com/demos/buttons/#third-party-routing-library

Fragment refs was proposed in reactjs/rfcs#97.